File: content/snippets/python/language_basics_exceptions_and_recovery.py
#!/usr/bin/env python3


def parse_port(value: str) -> int:
    return int(value)


def main() -> None:
    try:
        port = parse_port("not-a-number")
        print("port:", port)
    except ValueError as error:
        print("parse failed:", error)
        print("fallback port:", 8080)

    print("program continues")


if __name__ == "__main__":
    main()

File: content/snippets/typescript/language_basics_exceptions_and_recovery.ts
#!/usr/bin/env ts-node

import { setImmediate } from "node:timers";

function parsePort(value: string): number {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) {
    throw new Error("invalid port");
  }
  return parsed;
}

function main(): void {
  try {
    const port = parsePort("not-a-number");
    console.log("port:", port);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.log("parse failed:", message);
    console.log("fallback port:", 8080);
  }

  console.log("program continues");
}

main();
setImmediate(() => {});
